lib/ecto/type.ex

```diff
@@ -988,6 +991,14 @@ defmodule Ecto.Type do
     equal_map?(fun, Map.to_list(map1), map2)
   end
 
+  defp equal_map?(fun, [{key, nil} | tail], other_map) do
+    case Map.fetch(other_map, key) do
+      {:ok, nil} -> equal_map?(fun, tail, other_map)
```

Comment:
Pattern matching here is a bit more efficient:

```elixir
case other_map do
  %{^key => nil} -> equal_map?(fun, tail, other_map)
  _ -> false
end
```

You can also change the clause below if you want to. :)